fedimint-api/src/db/mod.rs

```diff
@@ -59,13 +61,105 @@ pub type PrefixIter<'a> = Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> +
 
 #[async_trait]
 pub trait IDatabase: Debug + Send + Sync {
-    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction;
+    async fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + Send + 'a>;
 }
 
-dyn_newtype_define! {
-    /// A handle to a type-erased database implementation
-    #[derive(Clone)]
-    pub Database(Arc<IDatabase>)
+#[derive(Clone, Debug)]
+pub struct Database(Arc<DatabaseInner<dyn IDatabase>>);
+
+// NOTE: `Db` is used instead of just `dyn IDatabase`
+// because it will impossible to construct otherwise
+#[derive(Debug)]
+struct DatabaseInner<Db: IDatabase + ?Sized> {
+    notifications: Notifications,
+    // once cell is used to solve cyclic initialization
+    // some modules need Database to be create
+    // and can provide decoders only after creation
+    module_decoders: OnceCell<ModuleDecoderRegistry>,
+    db: Db,
+}
+
+const WAIT_KEY_TIMEOUT: Duration = Duration::from_secs(60);
+
+impl<D: IDatabase + 'static> From<D> for Database {
+    fn from(db: D) -> Self {
+        Self::new(db)
+    }
+}
+
+impl Database {
+    pub fn new(db: impl IDatabase + 'static) -> Self {
+        let inner = DatabaseInner {
+            db,
+            notifications: Notifications::new(),
+            module_decoders: OnceCell::new(),
+        };
+
+        Self(Arc::new(inner))
+    }
+
+    pub fn register_module_decoders(&self, decoders: ModuleDecoderRegistry) {
+        self.0
+            .module_decoders
+            .set(decoders)
+            .expect("module decoders already registered");
+    }
+
+    // uses default decoders if not registered
+    fn decoders(&self) -> &ModuleDecoderRegistry {
+        static DEFAULT_MODULES: Lazy<ModuleDecoderRegistry> =
+            Lazy::new(ModuleDecoderRegistry::default);
+
+        self.0.module_decoders.get().unwrap_or(&DEFAULT_MODULES)
+    }
+
+    pub async fn begin_transaction(&self) -> DatabaseTransaction {
+        DatabaseTransaction::new(
+            self.0.db.begin_transaction().await,
+            self.decoders(),
+            &self.0.notifications,
+        )
+    }
+
+    async fn wait_key_inner(&self, key_bytes: &[u8]) -> Result<Vec<u8>> {
```

Comment:
What is the reasoning behind using key_bytes: &[u8] instead of the traits (Databasekey and DatabaseKeyPrefixConst) that get_value (in DatabaseTransaction) uses?

Reply:
Oh I see, its the inner function. Nevermind